### DISCUSSION TITLE
fpm optimization report generator

### DISCUSSION THREAD

**Discussion by ivan-pi** [fortran-lang/fpm#573] [Ideas]
Some compilers have the possibility to generate optimization reports like
* [-qopt-report](https://software.intel.com/content/www/us/en/develop/documentation/fortran-compiler-oneapi-dev-guide-and-reference/top/compiler-reference/compiler-options/compiler-option-details/optimization-report-options/qopt-report-qopt-report.html) (Intel)
* [-fopt-info](https://gcc.gnu.org/onlinedocs/gcc/Developer-Options.html) (GNU)

In the past I've always found it difficult to integrate these flags with my makefile and CMake build process. Normally too much information is output, which then needs to be filtered. Running the compiler directly is also cumbersome, if a subroutine depends on multiple other modules. During development I would like to limit optimization reports to a single module, or even single subroutine or function. 

I was envisioning something like:
```
fpm diagnose <module>:<submodule>:<subroutine or function> --compiler-flags=<...> 
```
would rebuild the module with the desired report flags and filter the output.

Perhaps someone else has a good idea how to embed something like this in the _edit, compile, run_ cycle when using fpm.

**Comment by LKedward** [comment]
Yes, I also would like to integrate optimization reporting into fpm somehow. I would probably write the report to a separate file for each translation unit which fpm can then easily interrogate based on module name.

We could add a `--report` or `--opt-info` flag to `fpm build` that would rebuild with the additional optimization report flags and then dump the relevant report to stdout:
```
$ fpm build --report [<module filter>] [build flags]
```

**Reply by ivan-pi** [reply]
That's a nice idea. I'm wondering if a two-step workflow, more like how [`gprof`](https://sourceware.org/binutils/docs/gprof/Compiling.html#Compiling) is used would be more suitable for this purpose.

For most projects I'm building currently, my commands are quite long due to various preprocessor or linkage of third-party libs.
Perhaps this becomes less of an issue, when preprocessing becomes supported.

**Reply by ivan-pi** [reply]
Speaking of gprof, it makes me wonder how gprof, gdb, valgrind, and the various Intel tools might be connected to fpm?
